In [2]:
import pandas as pd
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Ridge
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score

class DiabetesModeler:
    """A simple class to manage diabetes regression models."""
    
    def __init__(self):
        # Initialize the models we want to compare
        self.models = {
            "Linear Regression": LinearRegression(),
            "Ridge Regression": Ridge(),
            "Random Forest": RandomForestRegressor(n_estimators=100, random_state=42)
        }
        self.results = []

    def run_analysis(self):
        # 1. Load the built-in diabetes dataset 
        diabetes = datasets.load_diabetes()
        X, y = diabetes.data, diabetes.target

        # 2. Standard data division (80% train, 20% test) [cite: 28]
        # FIXED: Corrected 'test_size' argument here
        X_train, X_test, y_train, y_test = train_test_split(
            X, y, test_size=0.2, random_state=42
        )

        # 3. Train and evaluate each model
        for name, model in self.models.items():
            model.fit(X_train, y_train)
            predictions = model.predict(X_test)
            
            # Calculate metrics [cite: 30]
            mse = mean_squared_error(y_test, predictions)
            r2 = r2_score(y_test, predictions)
            
            self.results.append({
                "Model": name,
                "MSE": round(mse, 2),
                "R2 Score": round(r2, 4)
            })

    def display_results(self):
        # Convert results to a table for easy reading
        df = pd.DataFrame(self.results)
        print("--- Regression Model Performance ---")
        print(df)

# Main Execution
if __name__ == "__main__":
    my_project = DiabetesModeler()
    my_project.run_analysis()
    my_project.display_results()

--- Regression Model Performance ---
               Model      MSE  R2 Score
0  Linear Regression  2900.19    0.4526
1   Ridge Regression  3077.42    0.4192
2      Random Forest  2952.01    0.4428


**Best Performing model!**

After evaluating the three machine learning models, the Linear Regression model demonstrated the best overall performance. This conclusion is supported by two key metrics: Mean Squared Error (MSE) and the R^2 Score. The Linear Regression model achieved the lowest MSE of 2900.19, indicating its predictions were, on average, closer to the actual values than the other models. Furthermore, it produced the highest R^2 Score of 0.4526, meaning it explained approximately 45.26% of the variance in the data. While the Random Forest model was a close second (R^2 of 0.4428), the Linear Regression model remains the preferred choice here for its accuracy and simplicity on this dataset.